crates/codegen/parser/src/first_set.rs

```diff
@@ -0,0 +1,62 @@
+use crate::char_set::CharSet;
+
+pub struct FirstSet {
+    pub includes_epsilon: bool,
+    pub char_set: CharSet,
+}
+
+impl FirstSet {
+    pub fn new() -> Self {
+        Self {
+            includes_epsilon: false,
+            char_set: CharSet::empty(),
+        }
+    }
+
+    pub fn from_char_set(char_set: CharSet) -> Self {
+        Self {
+            includes_epsilon: false,
+            char_set,
+        }
+    }
+
+    pub fn single(c: char) -> Self {
+        Self {
+            includes_epsilon: false,
+            char_set: CharSet::single(c),
+        }
+    }
+
+    pub fn multiple<T: Iterator<Item = char>>(x: T) -> Self {
+        Self {
+            includes_epsilon: false,
+            char_set: x.fold(CharSet::empty(), |acc, c| acc.union(CharSet::single(c))),
+        }
+    }
+
+    pub fn epsilon() -> Self {
```

Comment:
IIUC, I suggest using `possibly_empty` or `can_be_empty` (or similar wording) instead of `epsilon`.
I realize it can be used in mathemantics settings, but I fear it can be confusing otherwise. We already have similar APIs that use `empty()` and `is_empty()`.

Reply:
epsilon is a basic and common term in the parsing literature and parsing theory. I'm not using it to be pretentious. It doesn't quite mean empty - that would be an empty set. It means that it derives an empty sequence.

Reply:
Of course. I didn't mean it is pretentious at all!
But the symbol epsilon `ε` is used in algebra/literature to indicate a certain meaning using a single letter. Here, we don't need to use a single letter. We can directly write its meaning in plain English in method and field names (in snake case of course 😅). This is similar to the `CharSet::empty()` we have now, as opposed to `CharSet::phi()`. The meaning is more understandable in English.

Reply:
As a compiler engineer, I know what a First Set is - it's a particular thing from this field of knowledge and study, as is a Pratt Parser, and as is an Epsilon Production, which is also used written out in prose, because the longhand description of what it means is too verbose. And pointless - we have a good term for what it is. In any case, to understand exactly what it means in the context of a First Set, and the implications thereof, requires some understanding of the theory.

But I'm curious - given you understand it does *not* mean empty set, what would you call it? And how would you not divorce it from the literature and its meaning in the specific context of First and Follow sets, and the algebra of parsing theory?

Reply:
Thanks for the reply. Maybe this suggestion isn't useful here.
Let's leave it as-is then.